src/common/dsp/effects/AudioInputEffect.cpp

```diff
@@ -0,0 +1,245 @@
+    : Effect(storage, fxdata, pd)
+{
+    effect_slot_type slotType = getSlotType(fxdata->fxslot);
+    if (storage && (slotType == a_insert_slot || slotType == b_insert_slot))
+    {
+        int scene = slotType == a_insert_slot ? 1 : 0;
+        for (int i = 0; i < N_OUTPUTS; i++)
+            sceneDataPtr[i] = storage->scenesOutputData.getSceneData(scene, i);
+    }
+}
+void AudioInputEffect::init_ctrltypes()
+{
+    Effect::init_ctrltypes();
+
+    // -----  Audio Input
+    fxdata->p[in_audio_input_channel].set_name("Channel");
+    fxdata->p[in_audio_input_channel].set_type(ct_percent_bipolar_stereo);
+    fxdata->p[in_audio_input_channel].posy_offset = 1;
+
+    fxdata->p[in_audio_input_pan].set_name("Pan");
+    fxdata->p[in_audio_input_pan].set_type(ct_percent_bipolar_stereo);
+    fxdata->p[in_audio_input_pan].posy_offset = 1;
+
+    fxdata->p[in_audio_input_level].set_name("Level");
+    fxdata->p[in_audio_input_level].set_type(ct_decibel_attenuation);
+    fxdata->p[in_audio_input_level].posy_offset = 1;
+
+    // -----  Effect Input
+    fxdata->p[in_effect_input_channel].set_name("Channel");
+    fxdata->p[in_effect_input_channel].set_type(ct_percent_bipolar_stereo);
+    fxdata->p[in_effect_input_channel].posy_offset = 3;
+
+    fxdata->p[in_effect_input_pan].set_name("Pan");
+    fxdata->p[in_effect_input_pan].set_type(ct_percent_bipolar_stereo);
+    fxdata->p[in_effect_input_pan].posy_offset = 3;
+
+    fxdata->p[in_effect_input_level].set_name("Level");
+    fxdata->p[in_effect_input_level].set_type(ct_decibel_attenuation);
+    fxdata->p[in_effect_input_level].posy_offset = 3;
+
+    // -----  Scene Input
+    effect_slot_type slot_type = getSlotType(fxdata->fxslot);
+    if (slot_type == a_insert_slot || slot_type == b_insert_slot)
+    {
+        fxdata->p[in_scene_input_channel].set_name("Channel");
+        fxdata->p[in_scene_input_channel].set_type(ct_percent_bipolar_stereo);
+        fxdata->p[in_scene_input_channel].posy_offset = 5;
+
+        fxdata->p[in_scene_input_pan].set_name("Pan");
+        fxdata->p[in_scene_input_pan].set_type(ct_percent_bipolar_stereo);
+        fxdata->p[in_scene_input_pan].posy_offset = 5;
+
+        fxdata->p[in_scene_input_level].set_name("Level");
+        fxdata->p[in_scene_input_level].set_type(ct_decibel_attenuation);
+        fxdata->p[in_scene_input_level].posy_offset = 5;
+    }
+
+    // -----  Output
+    fxdata->p[in_output_width].set_name("Width");
+    fxdata->p[in_output_width].set_type(ct_percent_bipolar_stereo);
+    fxdata->p[in_output_width].posy_offset = 7;
+
+    fxdata->p[in_output_mix].set_name("Mix");
+    fxdata->p[in_output_mix].set_type(ct_percent);
+    fxdata->p[in_output_mix].posy_offset = 7;
+}
+
+void AudioInputEffect::init_default_values()
+{
+    fxdata->p[in_audio_input_channel].val.f = 0.0; // p0
+    fxdata->p[in_audio_input_pan].val.f = 0.0;     // p1
+    fxdata->p[in_audio_input_level].val.f = -48.0; // p2
```

Comment:
-96, to match the other comments I left.

Reply:
Done